src/cast.ts

```diff
@@ -0,0 +1,96 @@
+import type { Field } from './index.js'
+import { decodeUTF8, uint8Array } from './text.js'
+
+/**
+ * https://github.com/vitessio/vitess/blame/v19.0.3/go/mysql/json/helpers.go#L86-L112
+ */
+
+export function cast(field: Field, value: any): any {
+  if (value == null) {
+    return value
+  }
+
+  if (isBigInt(field)) {
+    return value
+  }
+
+  if (isDateOrTime(field)) {
+    return value
+  }
+
+  if (isDecimal(field)) {
+    return value
+  }
+
+  if (isJson(field)) {
+    return JSON.parse(value)
+  }
+
+  if (isIntegral(field)) {
+    return parseInt(value, 10)
+  }
+
+  if (isFloat(field)) {
+    return parseFloat(value)
+  }
+
+  if (isBinary(field)) {
+    return uint8Array(value)
+  }
+
+  return decodeUTF8(value)
+}
+
+/**
+ * These are integral, but we want to leave the `BigInt` casting to the caller.
+ *
+ * https://github.com/planetscale/database-js/pull/90
+ */
+
+function isBigInt(field: Field) {
+  return field.type === 'INT64' || field.type === 'UINT64'
+}
+
+/**
+ * https://github.com/vitessio/vitess/blob/v19.0.3/go/sqltypes/type.go#L103-L106
+ */
+
+function isDateOrTime(field: Field) {
+  return field.type === 'DATETIME' || field.type === 'DATE' || field.type === 'TIMESTAMP' || field.type === 'TIME'
```

Comment:
I find it a bit odd that `isIntegral` for examples checks against a list of types with `includes` vs this and `isBigInt` use a hand written set of `||` statements. I'd say just pick one.